tuist/tuist
# Allow specifying multiple configurations within project manifests

Resolves https://github.com/tuist/tuist/issues/160

### Short description 📝

Xcode supports specifying multiple build configurations beyond the default "Debug" and "Release". Providing the ability to leverage multiple custom build configuration will allow Tuist's adoption by more projects.

### Solution 📦

Allow specifying multiple configurations from the project manifests. 

Note: The support and logic for dealing with multiple configurations has been in Tuist for some time as such it only required additional work to expose it in the Project manifests in a compatible fashion.

### Implementation 👩‍💻👨‍💻

> Detail in a checklist the steps that you took to implement the PR.

- [x] Expose multiple configurations within project manifest 
- [x] Update scheme generator to look up appropriate build configuration
- [x] Add scheme linter to ensure it only references valid configurations

### Test Plan 🛠

- Run `tuist generate` within `fixtures/ios_app_with_multi_configs`
- Verify the generated project contains the build configurations "Debug", "Beta" and "Release"
- Verify the generated project has a scheme "App-Beta" which uses the "Beta" configuration when performing the run action
- Verify that referencing a configuration within the scheme that isn't defined in the project flags a listing error 

### Questions / Discussion ❓

#### `CustomConfiguration`

To maintain backwards compatibility a new type was introduced to help associate a named build configuration (`BuildConfiguration`) with a `Configuration`. 

Multi config

```swift
let configurations: [CustomConfiguration] = [
    .debug(name: "Debug", xcconfig: "../ConfigurationFiles/Debug.xcconfig"),
    .release(name: "Beta", xcconfig: "../ConfigurationFiles/Beta.xcconfig"),
    .release(name: "Release", xcconfig: "../ConfigurationFiles/Release.xcconfig"),
]

let settings = Settings(configurations: configurations)
```

compatibility with previous debug / release: 

```swift
let debug = Configuration(xcconfig: "../ConfigurationFiles/Debug.xcconfig")
let release = Configuration(xcconfig: "../ConfigurationFiles/Release.xcconfig")

let settings = Settings(debug: debug, release: release)
```

This is the approach adopted by this PR, however there could be another we can follow.

Initially I attempted to update `Configuration` to include the name and variant however this would slightly change the behavior for existing users

```swift

public struct Configuration: Equatable, Codable {
    public let buildConfiguration: BuildConfiguration
    public let settings: [String: String]
    public let xcconfig: String?

    public init(buildConfiguration: BuildConfiguration = .debug, settings: [String: String] = [:], xcconfig: String? = nil) {
        self.buildConfiguration = buildConfiguration
        self.settings = settings
        self.xcconfig = xcconfig
    }

    public static func settings(_ settings: [String: String], xcconfig: String? = nil) -> Configuration {
        return Configuration(settings: settings, xcconfig: xcconfig)
    }
}

public extension Configuration {
    static func debug(name: String = "Debug", settings: [String: String] = [:], xcconfig: String? = nil) -> Configuration {
        return Configuration(buildConfiguration: .debug(name: name), settings: settings, xcconfig: xcconfig)
    }
    
    static func release(name: String = "Release", settings: [String: String] = [:], xcconfig: String? = nil) -> Configuration {
        return Configuration(buildConfiguration: .release(name: name), settings: settings, xcconfig: xcconfig)
    }
}
```

The change here would be any existing `Configuration` would need to get a default name and variant to maintain backwards compatibility, but that would result in 

```swift
let debug = Configuration(xcconfig: "../ConfigurationFiles/Debug.xcconfig")
let release = Configuration(xcconfig: "../ConfigurationFiles/Release.xcconfig") // debug under the hood due to defaults
let settings = Settings(debug: debug, release: release)
```

This could be mitigated internally in the `Settings` to re-create `Configuration` with the appropriate names and variants and only copying across any settings or xcconfig references.

```swift
public init(base: [String: String] = [:],
                debug: Configuration? = nil,
                release: Configuration? = nil,
                defaultSettings: DefaultSettings = .recommended) {
        configurations = [
            .debug(settings: debug?.settings ?? [:], xcconfig: debug?.xcconfig),
            .release(settings: release?.settings ?? [:], xcconfig: release?.xcconfig)
        ]
        self.base = base
        self.defaultSettings = defaultSettings
    }
```

The second approach reduces the types (and possibly any confusion that comes with it) however does open up some loopholes when using the default release / debug api. 

#### Scheme usage 

The scheme use can do with some evaluation: to ensure backwards compatibility the type `BuildConfiguration` does this work well at the point of use?

```swift
let betaScheme = Scheme(name: "App-Beta",
                         shared: true,
                         buildAction: BuildAction(targets: ["App"]),
                         runAction: RunAction(config: .release(name: "Beta"), executable: "App"))
```

vs. perhaps using a `String` to identify the configuration by name?

```swift
let betaScheme = Scheme(name: "App-Beta",
                         shared: true,
                         buildAction: BuildAction(targets: ["App"]),
                         runAction: RunAction(configurationName: "Beta", executable: "App"))
```

The linter needs to ensure a matching configuration exists either way.
